### ISSUE TITLE
[Proposal] IClusterClient.Connect retry could simplify user code

### ISSUE THREAD

**Issue by aarondandy** [dotnet/orleans#3983]
As a newcomer to Orleans I was surprised to learn that `Connect` does not Just Work, especially considering I was in my first already messy HelloWorld attempt. I expected to have something like the following after [reading the docs](http://dotnet.github.io/orleans/Documentation/Getting-Started-With-Orleans/Developing-a-Client.html) :

```csharp
static async Task JustConnect()
{
    var config = ClientConfiguration.LocalhostSilo();
    var builder = new ClientBuilder()
        .ConfigureApplicationParts(parts => parts.AddApplicationPart(typeof(ICorn).Assembly)
            .WithReferences())
        .ConfigureLogging(logging => logging.AddConsole())
        .UseConfiguration(config);
    ClusterClient = builder.Build();
    await ClusterClient.Connect();
}
```

Because my silo application takes a few seconds to start up the first few Connect calls are mostly guaranteed to fail. The first thing I attempted to do about it was to wrap the call to `Connect` in a `do {} while (true);` which provided me with an `InvalidOperationException` and a message of "A prior connection attempt failed. This instance must be disposed." Okay, I can build a new one! So I then moved my call to `.Build()` within my loop. That's a paddlin' too. "Build may only be called once per ClientBuilder instance." Okay, fine I guess. So I move my builder into the loop too and get:

```csharp
static async Task KeepTryingToConnect()
{
    var config = ClientConfiguration.LocalhostSilo();

    do
    {
        try
        {
            var builder = new ClientBuilder()
                .ConfigureApplicationParts(parts => parts.AddApplicationPart(typeof(ICorn).Assembly)
                    .WithReferences())
                .ConfigureLogging(logging => logging.AddConsole())
                .UseConfiguration(config);
            ClusterClient = builder.Build();
            await ClusterClient.Connect();
            break;
        }
        catch (SiloUnavailableException)
        {
            ClusterClient?.Dispose();
        }
    }
    while (true);
}
```

If some kind of retry could be built into `Connect` I think it could improve the new user experience a bit. I'm sure established projects and users also use a very similar pattern so maybe it can help others too.

**Comment by veikkoeeva** [comment]
@aarondandy You are right, this bites. See also https://github.com/dotnet/orleans/issues/3830.

I would be weary of building in such logic since it feels the shape of the retry function and it's characteristics (how many times, what kind of time interals between calls) can vary between application and business domain. Basically I'm advocating breaking immmdiately and providing instructions to use a circuit breaker such as  [Polly](https://github.com/App-vNext/Polly).

Maybe one way of doing this would be to provide an extension method that takes a circuit breaker function and returns a working client and tries the connection using the breaker until it succeeds or the circuit breaks. However, this can fail at runtime too (see #3830) and I believe I'd like to be notified it immediately and take appropriate action, such a have a retry. For this one has to have a container of some sort for the configuration so a new client can be built when the old one has broken.

This thinking comes from industrial automation where hidden buffers or hidden delays can topple down large, dependent systems.

**Comment by sergeybykov** [comment]
Resolved via #4161.
